---
title: "BLAST results summarized"
author: "Gavin Douglas"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  html_document:
    code_folding: hide
    theme: cerulean
    toc: true
    toc_float: true
---

```{r setup}
suppressPackageStartupMessages(library(ComplexHeatmap))
suppressPackageStartupMessages(library(circlize))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(DT))
suppressPackageStartupMessages(library(ggbeeswarm))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(kableExtra))
suppressPackageStartupMessages(library(knitr))
suppressPackageStartupMessages(library(reshape2))
```

Summary of BLAST results:

1. Numbers of HGT events by taxonomic level. 
2. Numbers of genes/bp per event by taxonomic level.  
3. COG enrichment results.  
4. Other genomic annotation enrichment results.  

# Putative HGT events by taxonomic level

```{r read_BLAST_nums}
num_comparisons <- read.table("/mfs/gdouglas/projects/water_mags/water_mag_analysis/summary_files/num_comparisons_per_inter.level.tsv.gz",
                          header = TRUE, sep = '\t', stringsAsFactors = FALSE, row.names = 1)

tally_tab <- read.table("/mfs/gdouglas/projects/water_mags/water_mag_analysis/summary_files/BLAST_hit_tally_table.tsv.gz",
                        header = TRUE, sep = '\t', stringsAsFactors = FALSE)
colnames(tally_tab)[which(colnames(tally_tab) == "Level")] <- "Inter-level"

# Ignore Species and Strain levels.
tally_tab <- tally_tab[which(! tally_tab$`Inter-level` %in% c('Species', 'Strain')), ]
num_comparisons <- num_comparisons[which(! rownames(num_comparisons) %in% c('Species', 'Strain')), , drop = FALSE]
```

# Count breakdowns

Note that the taxonomic levels represent putative HGT events between taxa that differ at that level. I.e., the tally at the "Family" level represents different taxa of the same Order (i.e., inter-family), but that differ at the family level.

See the _Ratio_ table to see how the relative numbers of hits in each category varies based on the >=95-99 vs the >=99% identity category.

Also note that a lot of hits between different genomes involved the same genes in different pairwise comparisons, which you can see from the difference between the "Num_total_genes" and "Num_unique_genes".

## Count tables {.tabset}

### Identity >= 95% and < 99%

```{r tally_table_95}
tally_tab_95 <- tally_tab[which(tally_tab$Identity == "Identity >= 95% and < 99%"), ]
kable(tally_tab_95, row.names = FALSE) %>%
   kable_styling(full_width = FALSE)
```

### Identity >= 99%

```{r tally_table_99}
tally_tab_99 <- tally_tab[which(tally_tab$Identity == "Identity >= 99%"), ]
kable(tally_tab_99, row.names = FALSE) %>%
   kable_styling(full_width = FALSE)
```

### Ratio: Identity >= 95% and < 99% vs. Identity >= 99%

```{r tally_table_ratio}
tally_tab_ratio <- tally_tab_95
tally_tab_ratio$Num_hits <- round(tally_tab_ratio$Num_hits / tally_tab_99$Num_hits, 2)
tally_tab_ratio$Num_total_genes <- round(tally_tab_ratio$Num_total_genes / tally_tab_99$Num_total_genes, 2)
tally_tab_ratio$Num_unique_genes <- round(tally_tab_ratio$Num_unique_genes / tally_tab_99$Num_unique_genes, 2)
tally_tab_ratio$Num_unique_clusters <- round(tally_tab_ratio$Num_unique_clusters / tally_tab_99$Num_unique_clusters, 2)

tally_tab_ratio$Identity <- "95% vs 99%"

kable(tally_tab_ratio, row.names = FALSE) %>%
   kable_styling(full_width = FALSE)
```

## Count tables normalized by number of comparisons {.tabset}

Comparing the raw counts can be misleading, because you need to know how many taxa that were compared differed at each taxonomic level to know if there was an enrichment.

Below are the same count tables as above, but **divided by the number of pairwise comparisons where the genomes differed at that taxonomic level**.

### Identity >= 95% and < 99%

```{r tally_table_95_divided_total}
tally_tab_95_norm <- tally_tab_95
tally_tab_95_norm$Num_hits <- round(tally_tab_95_norm$Num_hits / num_comparisons, 6)
tally_tab_95_norm$Num_total_genes <- round(tally_tab_95_norm$Num_total_genes / num_comparisons, 6)
tally_tab_95_norm$Num_unique_genes <- round(tally_tab_95_norm$Num_unique_genes / num_comparisons, 6)
tally_tab_95_norm$Num_unique_clusters <- round(tally_tab_95_norm$Num_unique_clusters / num_comparisons, 6)
kable(tally_tab_95_norm) %>%
   kable_styling(full_width = FALSE)
```

### Identity >= 99%

```{r tally_table_99_divided_total}
tally_tab_99_norm <- tally_tab_99
tally_tab_99_norm$Num_hits <- round(tally_tab_99_norm$Num_hits / num_comparisons, 6)
tally_tab_99_norm$Num_total_genes <- round(tally_tab_99_norm$Num_total_genes / num_comparisons, 6)
tally_tab_99_norm$Num_unique_genes <- round(tally_tab_99_norm$Num_unique_genes / num_comparisons, 6)
tally_tab_99_norm$Num_unique_clusters <- round(tally_tab_99_norm$Num_unique_clusters / num_comparisons, 6)
kable(tally_tab_99_norm) %>%
   kable_styling(full_width = FALSE)
```

# Odd's ratios by COG category {.tabset}

COG category enrichment was performed based on the functional annotation of each CD-HIT cluster. All clusters were categorized as either being found in a BLAST hit region for a given intra-taxon level or not (which should help with the issue of the same gene being hit numerous times at the same level, as they would only be counted once). The background was taken to be COG category frequency across all gene family clusters across all tested genomes.

Point colour represents FDR value:
  
* Dark red: < 0.05 
* Pink: < 0.20 
* White >= 0.20 

Although correlated, the enrichments based on the two different identity cut-offs do substantially differ. Does this give us any insight into which functions are most commonly transferred vs those that are most adaptive? Or would that be inappropriate to consider here?

```{r read_in}
COG_enrich <- read.table("/mfs/gdouglas/projects/water_mags/water_mag_analysis/summary_files/BLAST_COG_category_enrichment_no.unannot.tsv.gz",
                         header = TRUE, sep = "\t", stringsAsFactors = FALSE)

COG_enrich <- COG_enrich[which(! COG_enrich$taxon %in% c('Species', 'Strain')), ]

COG_enrich$taxon <- factor(COG_enrich$taxon, levels = c("Genus", "Family", "Order", "Class", "Phylum", "Domain"))

COG_enrich[which(COG_enrich$identity == "95"), "identity"] <- "id >= 95% and < 99%"
COG_enrich[which(COG_enrich$identity == "99"), "identity"] <- "id >= 99%"

```

```{r COG_category_breakdown, results='asis', echo=FALSE}

for (COG_category in unique(COG_enrich$category)) {

  COG_enrich_subset <- COG_enrich[which(COG_enrich$category == COG_category), ]
  
  cat('##', COG_enrich_subset$descrip[1], ' \n')

  COG_enrich_subset$sig_col <- "white"
  COG_enrich_subset$sig_col[which(COG_enrich_subset$fdr < 0.20)] <- "pink"
  COG_enrich_subset$sig_col[which(COG_enrich_subset$fdr < 0.05)] <- "dark red"
  
  print(ggplot(data = COG_enrich_subset, aes(x = log2(OR + 0.1), y = taxon)) +
    geom_point(fill = COG_enrich_subset$sig_col, size = 2, shape = 21) +
    facet_wrap(. ~ identity) +
    ggtitle(COG_enrich_subset$descrip[1]) +
    #scale_x_continuous(trans=scales::pseudo_log_trans(base = 10)) +
    geom_vline(xintercept = 0, linetype = 2, color = "blue", linewidth = 1) +
    ylab("Inter-taxon level") +
    theme_bw())
  
  cat('\n\n')

}
```

# COG category result heatmap

Viewing all the data as a heatmap is a lot easier to process

```{r COG_heatmap, fig.height=15, fig.width=15}
unique_COG_categories <- sort(unique(COG_enrich$descrip))
unique_taxa <- sort(unique(COG_enrich$taxon))

COG_ORs_95 <- data.frame(matrix(NA, nrow = length(unique_COG_categories), ncol = 6))
rownames(COG_ORs_95) <- unique_COG_categories
colnames(COG_ORs_95) <- unique_taxa

COG_ORs_99 <- COG_ORs_95

for (COG_cat in unique_COG_categories) {

  for (taxon in unique_taxa) {

    row_95 <- COG_enrich[which(COG_enrich$descrip == COG_cat & COG_enrich$taxon == taxon & COG_enrich$identity == "id >= 95% and < 99%"), ]
    if (nrow(row_95) == 1) {
      COG_ORs_95[COG_cat, taxon] <- row_95$OR
    }
    
    row_99 <- COG_enrich[which(COG_enrich$descrip == COG_cat & COG_enrich$taxon == taxon & COG_enrich$identity == "id >= 99%"), ]
    if (nrow(row_99) == 1) {
      COG_ORs_99[COG_cat, taxon] <- row_99$OR
    }
    
  }
}

colnames(COG_ORs_95) <- paste(colnames(COG_ORs_95), "95")
colnames(COG_ORs_99) <- paste(colnames(COG_ORs_99), "99")

COG_ORs <- cbind(COG_ORs_95, COG_ORs_99)

COG_ORs <- log2(COG_ORs)

COG_ORs_str <- format(round(COG_ORs, 3), nsmall = 3)
COG_ORs_str[COG_ORs_str == "    NA"] <- ""

draw(column_title = "",
     Heatmap(matrix = as.matrix(COG_ORs),
             name = "log2(OR)",
             row_names_side = "left",
             row_labels = rownames(COG_ORs),
             column_labels = c(unique_taxa, unique_taxa),
             column_split = c(rep("id >= 95% and < 99%", 6), rep("id >= 99%", 6)),
             cluster_rows = TRUE,
             cluster_columns = FALSE,
             column_gap = unit(5, "mm"),
             column_names_rot = 45,
             col = colorRamp2(c(-3, 0, 3), c("slateblue1","white", "red")),
             cell_fun = function(j, i, x, y, width, height, fill) {
               grid.text(COG_ORs_str[i, j], x, y, gp = gpar(fontsize = 10, fontface="bold"))
             }))
```

# Enrichments by other genomic annotations

```{r other_annot, fig.height=8, fig.width=15}
annot_enrich <- read.table("/mfs/gdouglas/projects/water_mags/water_mag_analysis/summary_files/BLAST_HGT_genomic_annot_enrichments.tsv.gz",
                         header = TRUE, sep = "\t", stringsAsFactors = FALSE)

annot_enrich$fdr <- p.adjust(annot_enrich$P, 'fdr')

annot_enrich[which(annot_enrich$fdr >= 0.05), 'OR'] <- NA

annot_enrich$log2OR <- log2(annot_enrich$OR)

annot_enrich_wide <- reshape2::dcast(data = annot_enrich, formula = annot ~ category, value.var = 'log2OR')

rownames(annot_enrich_wide) <- annot_enrich_wide$annot

annot_enrich_wide <- annot_enrich_wide[, c('Genus_95', 'Family_95', 'Order_95', 'Class_95', 'Phylum_95', 'Domain_95',
                                           'Genus_99', 'Family_99', 'Order_99', 'Class_99', 'Phylum_99', 'Domain_99')]

annot_enrich_wide_str <- format(round(annot_enrich_wide, 3), nsmall = 3)
annot_enrich_wide_str[annot_enrich_wide_str == "    NA"] <- ""
annot_enrich_wide_str[annot_enrich_wide_str == "   NA"] <- ""
annot_enrich_wide_str[annot_enrich_wide_str == "NA"] <- ""
annot_enrich_wide_str[annot_enrich_wide_str == "  NA"] <- ""

draw(column_title = "",
     Heatmap(matrix = as.matrix(annot_enrich_wide),
             name = "log2(OR)",
             row_names_side = "left",
             row_labels = rownames(annot_enrich_wide),
             column_labels = c(unique_taxa, unique_taxa),
             column_split = c(rep("id >= 95% and < 99%", 6), rep("id >= 99%", 6)),
             cluster_rows = FALSE,
             cluster_columns = FALSE,
             column_gap = unit(5, "mm"),
             column_names_rot = 45,
             col = colorRamp2(c(-5, 0, 5), c("slateblue1","white", "red")),
             cell_fun = function(j, i, x, y, width, height, fill) {
               grid.text(annot_enrich_wide_str[i, j], x, y, gp = gpar(fontsize = 10, fontface="bold"))
             }))
```

# Number of gene hits per pairwise BLAST region {.tabset}

Clearly some strains are just highly similar over huge regions, unsurprisingly.

Perhaps the biggest point of interest is that the average number of genes per region is actually pretty stable at the genus level and higher, but that regions that are more divergent tend to be smaller (which would make sense if they were older).

## With outliers

```{r genes_per_region_w_outliers}
hit_counts_per_region <- readRDS("/mfs/gdouglas/projects/water_mags/water_mag_analysis/summary_files/per_hit_gene_counts.rds")

hit_counts_per_region <- hit_counts_per_region[which(! hit_counts_per_region$Level %in% c('Species', 'Strain')), ]

hit_counts_per_region$Level <- factor(hit_counts_per_region$Level, levels = c("Domain", "Phylum", "Class", "Order", "Family", "Genus"))

ggplot(data = hit_counts_per_region, aes(x = Num_covered_genes, y = Level)) +
  geom_boxplot() +
  theme_bw() +
  facet_wrap(. ~ Identity) +
  xlab("Number of genes at least 90% covered by hit region\n(Average of genes in overlapping region of both genomes)") +
  ylab("Inter-taxon level")
```


## Outliers removed

```{r genes_per_region_no_outliers}
ggplot(data = hit_counts_per_region, aes(x = Num_covered_genes, y = Level)) +
  geom_boxplot(outlier.shape = NA) +
  theme_bw() +
  facet_wrap(. ~ Identity) +
  xlab("Number of genes at least 90% covered by hit region\n(Average of genes in overlapping region of both genomes)") +
  ylab("Inter-taxon level") +
  scale_x_continuous(limits = c(0, 20))
```


# Session info {.tabset}

## Hide

## Show

```{r show_info}
sessionInfo()
```
